### PR TITLE
support epoch in hf pipeline

### PR DESCRIPTION
**Pull request by aireenmei**
# Description
* support multi epoch training in hf pipeline. Multi-epoch is already supported in other pipelines . b/482070325
* migrate data input validations from pyconfig_deprecated.py to types.py
* Fix mdformat for correct numbering and fix numbering in some docs 

# Tests
Test on a small dataset with 1 epoch, data run out after 21 steps: https://paste.googleplex.com/5202487038574592
Test on the same dataset with 2 epochs, data run out after 42 steps: https://paste.googleplex.com/5516188497084416
Also used this [script](https://paste.googleplex.com/6192803555180544) to test the effects of `.repeat()` and `.shuffle()`. The output is as expected.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by NuojCheng — APPROVED** [review]

**Comment by aireenmei** [comment]
@SamuelMarks could you review the change in `.pre-commit-config.yaml`?

File: src/MaxText/input_pipeline/_hf_data_processing.py [hunk 1]
```diff
@@ -61,8 +61,13 @@ def vision_sft_preprocessing_pipeline(
   else:
     batch_size = global_batch_size // jax.process_count()
 
-  if config.enable_data_shuffling:
+  if config.enable_data_shuffling and config.num_epoch > 1:
```
Comment (hengtaoguo) [inline src/MaxText/input_pipeline/_hf_data_processing.py:65]: Can we add some one-liner comments to these if-else?

Comment (aireenmei) [reply src/MaxText/input_pipeline/_hf_data_processing.py:65]: added
